design-documents/graph-ql/schema-development.md

```diff
@@ -0,0 +1,108 @@
+# [GraphQL][] Schema Development
+
+Before developing a [GraphQL schema](https://graphql.org/learn/schema/), we need to consider the following:
+
+- [GraphQL Ecosystem](#graphql-ecosystem)
+- [GraphQL Server](#graphql-server)
+- [Code-First vs. SDL](#code-first-vs-sdl)
+- [Code-First Frameworks](#code-first-frameworks)
+   - [TypeGraphQL](#typegraphql)
+   - [Nexus](#nexus)
+
+## GraphQL Ecosystem
+
+[GraphQL][] is [a specification](https://graphql.github.io/graphql-spec/) for which a good handful of implementations are written across various languages (e.g., [Rust](https://github.com/graphql-rust), [PHP](https://github.com/webonyx/graphql-php), [Python](https://graphene-python.org/), [.NET](https://github.com/graphql-dotnet/graphql-dotnet) and [JavaScript](https://www.npmjs.com/package/graphql)). There is no wrong choice between these implementations; however, when considering the ecosystem, there is a clear winner.
+
+In 2015, the [GraphQL][] spec was [first published](https://graphql.github.io/graphql-spec/July2015/) along with the reference implementation in JavaScript. It should be no surprise then, that today, JavaScript holds the strongest ecosystem of [GraphQL][] libraries, with [4,833 related npm packages](https://www.npmjs.com/search?q=graphql) at the time of this writing (10/2019). As such, the JavaScript ecosystem is a solid and safe choice, moving forward.
+
+## GraphQL Server
+
+Within the JavaScript ecosystem, there are [a few server choices available](https://www.npmtrends.com/apollo-server-vs-koa-graphql-vs-prisma), but the most popular and mature choice is [Apollo Server][]. Optionally, [`graphql-yoga`](https://www.npmjs.com/search?q=graphql-yoga) uses [Apollo Server][] under the hood.
+
+## Code-First vs. SDL
+
+As [previously mentioned](#graphql-ecosystem), there are myriad JavaScript libraries available to assist in the creation of a [GraphQL][] server, but before we explore them, we should consider the pros & cons of defining our schema as code-first vs. SDL. Many articles [have been written](#references) about this topic and there are excellent arguments on both sides, but code-first appears to be the direction things are headed.
+
+In this document, we will explore the code-first approach.
+
+> Instead of increasing the complexity of GraphQL server development with more tools, we should strive for a simpler development model. Ideally, one that lets developers leverage the programming language they're already using – [this is the idea of code-first](https://www.prisma.io/blog/the-problems-of-schema-first-graphql-development-x1mn4cb0tyl3?source=post_page-----cf0e50d5ccff----------------------#code-first-a-language-idiomatic-way-for-graphql-server-development).
+
+### Pros
```

Comment:
I think a lot of these pros are missing information.

> SDL is a generated artifact, not manually maintained.

Can you explain what benefits this provides?

> The shape of the API is closer to the data model

Is this a pro? One of the benefits of GraphQL is that it's easier to not leak your backend's architecture into client-side code.

> Better code reuse and type safety

Probably needs some examples. SDL describes the types, so it's not clear why this could not be used to generate TypeScript (or other) types

> Declarative

SDL looks pretty declarative to me 🤔. Can you explain this?

> Can split by features/modules

Example here would be helpful, how this can be done with code-first but not with SDL

> Scalable

Seems like this is a relatively new way of doing things, might be a bit premature to throw that label on 😄

Reply:
I was trying to keep the list simple, but apparently that's too confusing. I'll try to add some clarity to these points.

Reply:
Before I do, though, let me see if you like these responses:

1. The main benefit is that you define your schema & resolvers in the same place (in code) instead of going back & forth between SDL & code (i.e., resolvers, etc.).
1. I don't see how this would be leaking anything into client-side code, as all of this "code" is written on the back end. Are you suggesting the SDL would be developed on the front end?
1. You _can_ generate TypeScript from SDL via [GraphQL Codegen](https://graphql-code-generator.com/), but because you can't define resolvers in SDL, you still have to bounce between SDL & code to complete the models.
1. SDL is definitely declarative. I realize now that this "Pros" list implies things that SDL are not. That was not my intent, but I can remove this bullet point to make that more clear?
1. Again, didn't mean to imply any of these pros could not also be done with SDL. I don't see any reason you couldn't also do this with SDL. Do you have any insights here?
1. It's scalable in that it's easier to maintain without introducing inconsistencies between SDL & code. I don't think that's premature at all.

Reply:
- I nested the "SDL is a generated artifact" point under the single source of truth bullet point to emphasise why it's a pro.
- I also nested "Scalable" under the same bullet point.
- Removed the "can split..." bullet.
- Removed "declarative".

See https://github.com/magento/architecture/pull/332/commits/7419e6e5ddf81b47e2bd6eb087a6a4c725469cc2

Reply:
- I nested the "SDL is a generated artifact" point under the single source of truth bullet point to emphasise why it's a pro.
- I also nested "Scalable" under the same bullet point.
- Removed the "can split..." bullet.
- Removed "declarative".

See https://github.com/magento/architecture/pull/332/commits/7419e6e5ddf81b47e2bd6eb087a6a4c725469cc2